### PR TITLE
.Net: API Manifest Plugins

### PR DESCRIPTION
**Pull request by zengin**
### Motivation and Context

This PR contains main implementation part of #4662 and adds it into `Functions.OpenApi.Extensions` project. There were also couple of bug fixes in the mentioned PR, those will be handled in subsequent PRs.

API Manifest can describe capabilities of an AI plugin for a chat-based system among other things. For a full description of API Manifest, see the [API Manifest Specification](https://darrelmiller.github.io/api-manifest/draft-miller-api-manifest.html).

API Manifest refers to one or more OpenAPI documents as API dependencies and it selects a subset of operations from those documents to describe the capabilities of an AI plugin.

### Description

Semantic Kernel has a way of creating plugins from OpenAPI operations, but it has a couple of limitations that can be addressed with API Manifest:
1. It can only create functions from OpenAPI operations that are described in a single OpenAPI document.
2. It cannot reference multiple OpenAPI documents in a single plugin.

Not being able to slice OpenAPI operations is a major limitation, especially for large APIs such as Microsoft Graph, which has over 1000 operations.

Today, we can create a Semantic Kernel plugin from an OpenAPI file and one OpenAPI file corresponds to one plugin. We don't have a way of selecting a subset of operations.

```mermaid
graph TB

subgraph OpenAPI1
operation11
operation12
operation13
end

subgraph OpenAPI2
operation21
operation22
operation23
end

subgraph OpenAPIPlugin1
function11
function12
function13
end

subgraph OpenAPIPlugin2
function21
function22
function23
end

operation11 -.- function11
operation12 -.- function12
operation13 -.- function13

operation21 -.- function21
operation22 -.- function22
operation23 -.- function23

style OpenAPIPlugin1 fill:blue,color:white
style OpenAPIPlugin2 fill:blue,color:white

```

### Proposed Solution
After Semantic Kernel APIManifest support, we will have an APIManifest file, which will be used as a filter on OpenAPI files and it can reference more than one OpenAPI file. The filter can be used as a guide to select only certain operations to be converted into functions.

Assume that we want to create a single plugin with `function11` from `OpenAPI1` and `function21` and `function22` from `OpenAPI2`. We want to preserve the one-to-one mapping between the APIManifest and the plugin, e.g. between `ApiManifestX` and `PluginX`:

```mermaid
graph TB

subgraph OpenAPI1
operation11
operation12
operation13
end

subgraph OpenAPI2
operation21
operation22
operation23
end

subgraph APIManifestX
reference11
reference21
reference22
end

subgraph PluginX
function11
function21
function22
end

reference11 -.- operation11
reference21 -.- operation21
reference22 -.- operation22

function11 -.- reference11
function21 -.- reference21
function22 -.- reference22

style APIManifestX fill:green,color:white
style PluginX fill:blue,color:white

```

If we use existing OpenAPI to Plugin public methods after on-the-fly slicing, we can only achieve the following because there is an assumption that a single OpenAPI is converted into a single plugin, which breaks APIManifest to Plugin one-to-one mapping. A single `APIManifestX` file corresponds to two plugins, namely, `PluginX1` and `PluginX2`: 

```mermaid
graph TB

subgraph OpenAPI1
operation11
operation12
operation13
end

subgraph OpenAPI2
operation21
operation22
operation23
end

subgraph APIManifestX
reference11
reference21
reference22
end

subgraph OpenAPI1_Sliced
operation11_sliced[operation11]
end

subgraph OpenAPI2_Sliced
operation21_sliced[operation21]
operation22_sliced[operation22]
end

subgraph PluginX1
function11
end

subgraph PluginX2
function21
function22
end

reference11 -.- operation11
reference21 -.- operation21
reference22 -.- operation22

operation11 -.- operation11_sliced -.- function11
operation21 -.- operation21_sliced -.- function21
operation22 -.- operation22_sliced -.- function22

style APIManifestX fill:green,color:white
style PluginX1 fill:blue,color:white
style PluginX2 fill:blue,color:white

```

To avoid this, the implementation of regular OpenAPI plugin generation and API Manifest plugin generation meets at the function generation level, so reuse starts at `CreateRestApiFunction`.

The PR also adds a KernelSyntaxExample for `/me/messages` endpoint of Microsoft Graph.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/Functions/Functions.OpenApi/Extensions/OpenApiKernelExtensions.cs [hunk 1]
```diff
@@ -258,7 +258,7 @@ public static class OpenApiKernelExtensions
     /// <param name="documentUri">The URI of OpenAPI document.</param>
     /// <param name="loggerFactory">The logger factory.</param>
     /// <returns>An instance of <see cref="KernelFunctionFromPrompt"/> class.</returns>
-    private static KernelFunction CreateRestApiFunction(
+    internal static KernelFunction CreateRestApiFunction(
```
Comment (zengin) [inline dotnet/src/Functions/Functions.OpenApi/Extensions/OpenApiKernelExtensions.cs:261]: @SergeyMenshykh, I am not sure if you prefer these to be `public` or `internal` (and exposed to `Functions.OpenApi.Extensions` project via `InternalsVisibleTo` in `csproj`). I went with the latter, but please advise if you prefer a different approach.

File: dotnet/samples/KernelSyntaxExamples/Example79_ApiManifest.cs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+using Microsoft.SemanticKernel.Plugins.OpenApi;
+using Microsoft.SemanticKernel.Plugins.OpenApi.Extensions;
+using Xunit;
+using Xunit.Abstractions;
+namespace Examples;
+
+// This example shows how to use the ApiManifest based plugins
+public class Example79_ApiManifest : BaseTest
+{
+    private void WriteSampleHeadingToConsole(string pluginToTest, string functionToTest, KernelArguments? arguments, params string[] pluginsToLoad)
```
Comment (SergeyMenshykh) [inline dotnet/samples/KernelSyntaxExamples/Example79_ApiManifest.cs:22]: nit: move all private members to the bottom of the class to emphasize the actual example.

File: dotnet/src/Functions/Functions.OpenApi.Extensions/Extensions/OpenApiKernelExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+            foreach (var path in filteredOpenApiDocument.Paths)
+            {
+                var operations = OpenApiDocumentParser.CreateRestApiOperations(serverUrl, path.Key, path.Value);
+                foreach (RestApiOperation operation in operations)
+                {
+                    try
+                    {
+                        logger.LogTrace("Registering Rest function {0}.{1}", pluginName, operation.Id);
+                        functions.Add(Microsoft.SemanticKernel.Plugins.OpenApi.OpenApiKernelExtensions.CreateRestApiFunction(pluginName, runner, operation, executionParameters, new Uri(serverUrl), loggerFactory));
```
Comment (SergeyMenshykh) [inline dotnet/src/Functions/Functions.OpenApi.Extensions/Extensions/OpenApiKernelExtensions.cs:146]: Consider moving the 'Microsoft.SemanticKernel.Plugins.OpenApi.OpenApiKernelExtensions' namespace to the 'using' directive.

File: dotnet/src/Functions/Functions.OpenApi/Extensions/OpenApiKernelExtensions.cs (hunk 1 above)
Comment (SergeyMenshykh) [reply dotnet/src/Functions/Functions.OpenApi/Extensions/OpenApiKernelExtensions.cs:261]: Let's go with the internal option for now; we can revise it later if needed.

File: dotnet/src/Functions/Functions.OpenApi.Extensions/Extensions/OpenApiKernelExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    /// <summary>
+    /// Imports a plugin from an API manifest asynchronously.
+    /// </summary>
+    /// <param name="kernel">The kernel instance.</param>
+    /// <param name="pluginName">The name of the plugin.</param>
+    /// <param name="filePath">The file path of the API manifest.</param>
+    /// <param name="executionParameters">Optional execution parameters for the plugin.</param>
+    /// <param name="cancellationToken">Optional cancellation token.</param>
+    /// <returns>The imported plugin.</returns>
+    [Experimental("SKEXP0099")]
```
Comment (zengin) [inline dotnet/src/Functions/Functions.OpenApi.Extensions/Extensions/OpenApiKernelExtensions.cs:35]: @SergeyMenshykh, is there anything I need to be aware of picking the ID for this? Can I pick the next increment of SKEXP?

File: dotnet/src/Functions/Functions.OpenApi.Extensions/Extensions/OpenApiKernelExtensions.cs (hunk 3 above)
Comment (zengin) [reply dotnet/src/Functions/Functions.OpenApi.Extensions/Extensions/OpenApiKernelExtensions.cs]: I had to use fully qualified name due to name collision of `OpenApiKernelExtensions` in both namespaces. I renamed this file to `ApiManifestKernelExtensions`.

**Review by SergeyMenshykh — APPROVED** [review]

**Comment by zengin** [comment]
@dmytrostruk @markwallace-microsoft can I get one more review, please? Thanks!

**Review by markwallace-microsoft — APPROVED** [review]

**Pull request merged**